cpp/inc/bond/core/detail/any.h

```diff
@@ -0,0 +1,307 @@
+        {
+            using base = impl<T, (sizeof(T) <= sizeof(storage))>;
+
+            template <typename U = T, typename boost::enable_if<std::is_copy_assignable<U> >::type* = nullptr>
+            static void assign(any& x, const T& value)
+            {
+                *base::unsafe_cast(x) = value;
+            }
+
+            template <typename U = T, typename boost::disable_if<std::is_copy_assignable<U> >::type* = nullptr>
+            static void assign(any& x, const T& value)
+            {
+                // TODO: Cache allocated buffer and reuse.
+                base::destroy(x);
+                emplace(x, value);
+            }
+
+            static void emplace(any& x, const T& value)
+            {
+                base::emplace(x, value);
+                x._id = TypeId<T>::value;  // Update the id if emplace did not throw.
+            }
+
+            static functions make() BOND_NOEXCEPT
+            {
+                return
+                {
+                    base::destroy,
+                    [](any& x, const any& other) { assign(x, *base::unsafe_cast(other)); },
+                    [](any& x, const any& other) { emplace(x, *base::unsafe_cast(other)); },
+                    [](const any& x, const any& y) { return *base::unsafe_cast(x) == *base::unsafe_cast(y); }
+                };
+            }
+        };
+    };
+
+
+    const void* data() const BOND_NOEXCEPT
+    {
+        return &_storage;
+    }
+
+    void* data() BOND_NOEXCEPT
+    {
+        return &_storage;
+    }
+
+    template <typename T>
+    void emplace(const T& value)
+    {
+        functions::template table<T>::emplace(*this, value);
+        _functions = functions::template table<T>::make();
+    }
+
+    void emplace(const any& other)
+    {
+        if (!other.empty())
+        {
+            other._functions.emplace(*this, other);
+            _functions = other._functions;
+        }
+        else
+        {
+            _id = {};
+        }
+    }
+
+    template <typename T>
+    any& assign(const T& value)
+    {
+        if (!try_assign_same(value))
+        {
+            // TODO: Cache allocated buffer and reuse.
+            destroy();
+            emplace(value);
+        }
+
+        return *this;
+    }
+
+    template <typename T>
+    bool try_assign_same(const T& value)
+    {
+        if (_id == TypeId<T>::value)
+        {
+            functions::template table<T>::assign(*this, value);
+            return true;
+        }
+
+        return false;
+    }
+
+    bool try_assign_same(const any& other)
+    {
+        if (_id == other._id)
+        {
+            if (!empty())
+            {
+                _functions.assign(*this, other);
+            }
+
+            return true;
+        }
+
+        return false;
+    }
+
+    void destroy()
+    {
+        if (!empty())
+        {
+            _functions.destroy(*this);
+            _id = {};
+        }
+    }
+
+
+    std::uint32_t _id;
```

Comment:
`_storage` then `_functions` then `_id` to minimize slack space added for alignment?